Add bulk insert support

Aflynn50 opened the pull request:
SQLair currently supports insert expressions, but not bulk inserts. To insert multiple rows into a database with SQLair, users currently have to execute the same statement multiple times using a for loop or write a query that inserts a fixed number of rows by explicitly writing out each row with separate input expressions for every column.

Additionally, users may want to bulk insert where network latency needs to be minimised such as in Juju where the distributed database requires multiple trips.

Bulk inserts are supported using the same syntax currently used for INSERTS. If one or more of the types passed to `Query` is a slice instead of a simple value, then it becomes a bulk insert with any other values in the INSERT remaining constant.

For example:
```
type S struct {
    v string `db:"col1"`
}

type T struct {
   v string `db:"col4"`
}

s := sqlair.MustPrepare(
    "INSERT INTO t (col1, col2, col3, col4) VALUES ($S.*, $M.col2, "v3", $T.col4)",
    S{},
    sqlair.M{},
    T{},
)

db.Query(
    s,
    []S{{v: "v1.1"}, {v: "v1.2"}, {v: "v1.3"}},
    sqlair.M{"col2": "v2"},
    []T{{v: "v4.1"}, {v: "v4.2"}, {v: "v4.3"}},
)
```
The query sent to the DB from the above example will be:
```
INSERT INTO  t (col1, col2, col3, col4) VALUES (@sqlair_0, @sqlair_3, "v3", @sqalir_4), (@sqlair_1, @sqlair_3, "v3", @sqalir_5), (@sqlair_2, @sqlair_3, "v3", @sqalir_6)
```
Where the placeholders are:
Name | Value
---|---
sqlair_0 | v1.1
sqlair_1 | v1.2
sqlair_2 | v1.3
sqlair_3 | v2
sqlair_4 | v4.1
sqlair_5 | v4.2
sqlair_6 | v4.3

Note that SQLair will check that all the slices passed to Query for bulk INSERTS are of the same cardinality, else throw out an error.

Aflynn50 committed:
Add bulk insert support for asterisk and column insert expressions

Aflynn50 committed:
Add support for insert expressions with literals

letFunny commented on internal/typeinfo/valuelocator.go:
```diff
@@ -134,16 +185,52 @@ func (f *structField) ArgType() reflect.Type {
+				// mix then throw an error.
+				if i == 0 && val.IsZero() {
+					omit = true
+				} else if val.IsZero() && !omit || !val.IsZero() && omit {
```
`val.IsZero() != omit`

letFunny commented on internal/expr/parser.go:
```diff
@@ -922,6 +930,26 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+func (p *Parser) parseColumnsInsertValues() ([]memberAccessor, bool, error) {
+	cp := p.save()
 	parenCol := p.colNum()
 	parenLine := p.lineNum
```
These are redundant now.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -45,47 +45,78 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
+						firstBulkColumn = bc.inputName
+						numRows = len(bc.vals)
+					}
+				}
```
Draft suggestion and questions:
```go
// do we need another error for when first column is NOT bulk but second one is? Right now bulk will be set on the second one?
	if bc.bulk && !bulk {
		// first row.
		bulk = true
		firstBulkColumn = bc.inputName
		numRows = len(bc.vals)
		continue
	}

	if bulk {
		if bc.bulk && len(bc.vals) != numRows {
			return nil, mismatchedBulkLengthsError(firstBulkColumn, numRows, bc.inputName, len(bc.vals))
		}
		if !bc.bulk {
			// error because all columns have to either be bulk or not?
		}
	}
```

Aflynn50 replied:
We have any some columns bulk and some not, the ones that are not are just duplicated

Aflynn50 replied:
I'll swap the order of the if/else, makes sense to have the first one first

letFunny commented on internal/expr/parser.go:
```diff
@@ -435,6 +417,32 @@ func (p *Parser) skipString(s string) bool {
 	return false
 }
 
+// skipLiteralInList advances the parser to the next comma or closing bracket
+// skipping string literals, comments and matching sets of parentheses.
+func (p *Parser) skipLiteralInList() (bool, error) {
+	for p.pos < len(p.input) {
+		if ok, err := p.skipStringLiteral(); err != nil {
+			return false, err
+		} else if ok {
+			continue
+		}
+		if ok, err := p.skipEnclosedParentheses(); err != nil {
+			return false, err
+		} else if ok {
+			continue
+		}
+		if ok := p.skipComment(); ok {
+			continue
+		}
+
+		if p.char == ',' || p.char == ')' {
+			return true, nil
+		}
+		p.advanceChar()
+	}
+	return false, nil
+}
```
This looks like a specific case of the parseUntil function we discussed months ago, maybe it will be helpful now.

Aflynn50 replied:
True, I think we should add that if we add another similar function to this in the future. The `advanceToNextExpression` function where you originally suggested using a `parseUntil` function isn't quite such a good fit for it now, though feel free to bring a PR if you think it could be useful once this is in

letFunny replied:
Only problem with that is that I doubt we are going to remember to refactor it if we need to create another one in the future. If we added it now it wouldn't add more than 5 lines (passing in the closure to get ',' or ')' and return the error).

Aflynn50 replied:
I'd rather leave it as it is, this way the name of the function documents exactly what it does.

letFunny commented on internal/expr/parser.go:
```diff
@@ -942,6 +970,57 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+
+		nextItem = p.skipChar(',')
+		p.skipBlanks()
+		itemStart = p.pos
```
This can be moved to the top of the loop to remove L983 and L986 above and make it more uniform.

letFunny commented on internal/expr/parser.go:
```diff
@@ -942,6 +970,57 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+			return vs, true, nil
+		}
+
+		nextItem = p.skipChar(',')
```
I think normally you would do:
```go
if !p.skipChar(',') {
    break
}
```
and a for loop without condition on top.

letFunny reviewed (commented):
Skimmed through it, looks good. Waiting for the proper review.

Aflynn50 committed:
Add comments and tidy up

SimonRichardson commented:
Out of curiosity more than anything. From the PR description, the SQL sent to the driver is:

```
INSERT INTO  t (col1, col2, col3, col4) VALUES (@sqlair_0, @sqlair_1, "v3", @sqalir_2), (@sqlair_3, @sqlair_4, "v3", @sqalir_5), (@sqlair_6, @sqlair_7, "v3", @sqalir_8)
```

Yet, `col2`, will always be a map for a constant value, so the query _could_ be turned into?

```
INSERT INTO  t (col1, col2, col3, col4) VALUES (@sqlair_0, @sqlair_1, "v3", @sqalir_2), (@sqlair_3, @sqlair_1, "v3", @sqalir_4), (@sqlair_5, @sqlair_1, "v3", @sqalir_6)
```

SimonRichardson reviewed (commented):
I'd love to see a full end to end test added to [package_test.go](https://github.com/canonical/sqlair/blob/main/package_test.go) that baked in the bulk insert changes.

I think the more common approach that I regularly want to use bulk insert in sqlair for is the following:

```
type S struct {
    a string `db:"col1"`
    b string `db:"col2"`
    c string `db:"col3"`
    d string `db:"col4"`
}

s := sqlair.MustPrepare(
    "INSERT INTO t (col1, col2, col3, col4) VALUES ($S.*)",
    S{},
)

db.Query(
    s,
    []S{
        {a: "a1", b: "b1", c: "c1", d: "d1"}, 
        {a: "a2", b: "b2", c: "c2", d: "d2"},
    },
)
```

Name | Value
--- | ---
sqlair_0 | "a1"
sqlair_1 | "b1"
sqlair_2 | "c1"
sqlair_3 | "d1"
sqlair_4 | "a2"
sqlair_5 | "b2"
sqlair_6 | "c2"
sqlair_7 | "d2"

manadart commented on internal/expr/parser.go:
```diff
@@ -922,16 +930,36 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+	return nil, false, nil
+}
+
+// parseColumnsInsertValues parses insert expressions that are value on the
```
"that are _values_ ..."?

manadart commented on internal/expr/parser.go:
```diff
@@ -942,6 +970,59 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+	return sources, true, nil
+}
+
+// parseBasicInsertValues parses insert expressions that are value on the
```
Same here.

manadart commented on internal/typeinfo/valuelocator.go:
```diff
@@ -32,9 +32,32 @@ type Input interface {
+	// Vals contains the query parameters.
+	Vals []any
+	// Omit indicates if the params have an omitempty flag set and should be
+	// ommited from input.
```
"omitted".

manadart commented on internal/typeinfo/valuelocator.go:
```diff
@@ -32,9 +32,32 @@ type Input interface {
+}
+
+// Params contains query parameters and metadata generated from an input value.
+type Params struct {
```
Good extraction.

manadart commented on internal/expr/bindinputs.go:
```diff
@@ -45,47 +45,74 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
-					return nil, fmt.Errorf("internal error: types in insert expressions cannot return multiple values")
+				bc.markArgUsed(argTypeUsed)
+
+				if bc.bulk {
```
These additions take the length of this method over 100 lines. We should seek ways to make it more terse.

letFunny replied:
This first part of getting the columns and the inputs seems like a good fit.

Aflynn50 replied:
I've refactored this method into a new `queryBuilder`. Then used the visitor pattern with the `typedExpr` to call the add methods on it.

manadart reviewed (commented):
Some comments from a partial review.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -45,47 +45,74 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
+			var boundColumns []*boundColumn
+			var columnNames []string
+			bulk := false
+			numRows := 1
```
I think we should have a comment here explaining what / how this works. Maybe combine it with the comment below:
```
// bulk detects whether we want to do a bulk insert (numRows >= 1). If that is the case, firstBulkColumn stores...
```
What do you think? Because right now I feel that you need to read the whole thing to get an idea of how it works. Or maybe this is alleviated completely by the method refactor manadart@ suggested.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +185,115 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
+	if bc.argType != nil {
+		argUsed[bc.argType] = true
+	}
+}
```
This method might be a little overkill seeing it is only used once.

Aflynn50 replied:
Fair, I've inlined it

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -45,47 +45,74 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
+						bulk = true
+						firstBulkColumn = bc.inputName
+						numRows = len(bc.vals)
+					} else if len(bc.vals) != numRows {
```
[Nit]: If this condition is relying on bulk == true, maybe it is better to put it explicitly or even as a separate if conditional, else you have to keep track of previous conditionals in your head.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -132,7 +159,14 @@ type typedInputExpr struct {
 }
 
-// insertColumn stores information about a single column of a row in an INSERT
+// typedColumn represents a column and input locator in an insert statement.
```
Can you add a `TODO` here or in the `columnAccessor` stating that we need to do a next to unify the column types?

Aflynn50 replied:
I've added a TODO to the `parseInsertColumns` to say that it needs to return the correct column type. I'm not sure we want to unify the parser column types and the "typed" column types.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +185,115 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
+		return nil, fmt.Errorf("internal error: types in insert expressions cannot return multiple values")
+	}
+	if params.Omit {
+		if ic.explicit {
```
[Nit]: combine with &&

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +185,115 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
+	argType reflect.Type
+	// inputName is the type name of the input parameter.
+	inputName string
+	// literal set if the value to insert is a literal.
```
```suggestion
	// literal is set if the value to insert is a literal.
```

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +185,115 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
+
+// boundColumn represents a column in an insert statement along with the values
+// to be inserted into it.
+type boundColumn struct {
```
Nice change! it unifies all the omit and bulk logic into one type.

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +185,115 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
+	name := "sqlair_" + strconv.Itoa(inputCount)
+	if !bc.bulk {
+		return "@" + name, sql.Named(name, bc.vals[0]), false, nil
+	}
```
Do we need this extra check? or can we interpret regular inserts as bulk or row == 1?

Aflynn50 replied:
Changed, I've made this whole method a switch case on len(bc.vals) instead

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -170,18 +309,24 @@ type sqlBuilder struct {
+func (b *sqlBuilder) writeInsert(columns []string, rows [][]string) {
 	// Write out the columns.
 	b.buf.WriteString("(")
 	b.writeCommaSeperatedList(columns, func(_ int, column string) string {
```
There is a typo in `Seperated`.

Aflynn50 replied:
Good spot, changed

letFunny commented on internal/expr/bindinputs.go:
```diff
@@ -239,3 +384,20 @@ func markerIndex(s string) (int, bool) {
 }
+
+func mismatchedBulkLengthsError(firstBulkColumn string, expectedNumRows int, badTypeName string, badRowLength int) error {
+	return fmt.Errorf("different slices sizes in bulk insert: slice of %q has length %d but slice of %q has length %d",
```
[Nit]: `slices sizes` is a little bit weird to read. What about `expected slices of same size in bulk insert: slice ....`?

letFunny commented on internal/expr/bindtypes.go:
```diff
@@ -259,6 +259,44 @@ func (e *columnsInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (tie any, err er
 	return &typedInsertExpr{insertColumns: cols}, nil
 }
 
+// basicInsertExpr is an input expression occurring within an INSERT statement
+// that consists of columns on the left and type accessors or literals on the
+// right. Unlike the columnInsertExpr, the values on the right are independent
+// of the columns on the left and are matched by position rather than by name.
+// e.g. (col1, col2, col3) VALUES ($M.key, "literal value", $T.value).
```
10/10 explanation.

letFunny commented on internal/expr/bindtypes.go:
```diff
@@ -259,6 +259,44 @@ func (e *columnsInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (tie any, err er
+		}
+	}()
+	if len(e.columns) != len(e.sources) {
+		return nil, fmt.Errorf("mismatched number of columns and values (%d!=%d)", len(e.columns), len(e.sources))
```
[Nit]: I think normally we use a colon before the inequality in the error messages, something like `mismatched number of columns and values: %d != %d`.

Aflynn50 replied:
Changed, and added tests for this error

letFunny commented on internal/expr/bindtypes.go:
```diff
@@ -378,6 +416,50 @@ func (e *outputExpr) bindTypes(argInfo typeinfo.ArgInfo) (te any, err error) {
+// valueAccessor defines an accessor that can be used to generate a typedColumn
+// with the given column name.
+type valueAccessor interface {
+	typedColumn(argInfo typeinfo.ArgInfo, columnName string) (typedColumn, error)
```
If I am reading this correctly, `typedColumn` is only used for insert columns? (context: memberAccessor below returns a `newInsertColumn`.

Aflynn50 replied:
And literals

letFunny replied:
literals in insert expressions, right?

Aflynn50 replied:
Yea, right now the valueAccessor is implemented by `memberAccessor` and `literal` the typedColumn interface is implemented by `literalColumn` and `insertColumn`.

Its not the best abstraction, but I was struggling to find a nice way to do literals.

letFunny commented on internal/expr/bindtypes.go:
```diff
@@ -259,6 +259,44 @@ func (e *columnsInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (tie any, err er
+
+// String returns a text representation for debugging and testing purposes.
+func (e *basicInsertExpr) String() string {
+	return fmt.Sprintf("RenamingInsert[%v %v]", e.columns, e.sources)
```
Maybe it would be useful to use commas to separate the sources, if you have an expression like `1 + 2 + 3` or even two words it might look like two separate sources instead of one.

Aflynn50 replied:
True, its a bit confusing, but this is only for our tests and will make the diff even worse for this PR.

letFunny commented on internal/expr/expr_test.go:
```diff
@@ -655,6 +668,70 @@ AND z = @sqlair_0 -- The line with $Person.id on it
+	expectedParams: []any{"Al", "const map", "const street", "Albert", "const map", "const street"},
+	expectedSQL:    `INSERT INTO person (name, mapKey, street) VALUES (@sqlair_0, @sqlair_1, @sqlair_2), (@sqlair_3, @sqlair_4, @sqlair_5)`,
+}, {
+	summary:        "bulk insert with omit empty struct not inserted at all",
```
Can you add a test case for a slice of `OmitEmptyID` where they are not omitted? And another one with mixed values, zero and non-zero?

Aflynn50 replied:
Added

letFunny commented on internal/expr/expr_test.go:
```diff
@@ -655,6 +668,70 @@ AND z = @sqlair_0 -- The line with $Person.id on it
+	expectedParams: []any{1, "val", "Church Road"},
+	expectedSQL:    `INSERT INTO person (id, key, street) VALUES (@sqlair_0, @sqlair_1, @sqlair_2)`,
+}, {
+	summary:        "bulk insert with explicit columns and literal",
```
This test is the same as the one below, right?

Aflynn50 replied:
Yea pretty much, the second is a more complex case of the first, though I'm inclined to keep both, because these also serve as examples to people looking at the library. 

I've removed the third insert value from the first one so that it at least tests a literal at the end.

letFunny commented on internal/expr/parser.go:
```diff
@@ -922,16 +930,36 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+		return &columnsInsertExpr{columns: columns, sources: sources, raw: p.input[cp.pos:p.pos]}, true, nil
+	}
+
+	if sources, ok, err := p.parseBasicInsertValues(); err != nil {
```
It is a little bit weird that `parseColumnsInsertExpr` has a `parseColumnsInsertValues` and a `parseBasicInsertValues`, it seems like the first one matches the intent and the second one does not belong here naming-wise.

Aflynn50 replied:
Yea, I've just redone this whole section, there was a load of redundancy and confusion there

letFunny commented on internal/expr/parser.go:
```diff
@@ -922,16 +930,36 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+}
+
+// parseColumnsInsertValues parses insert expressions that are value on the
+// right hand side of a columns insert expression.
```
Can you state what the expression is instead of `columns insert expression`. I find that confusing as the outer parse function is named the same and there is no clear indication what the difference is.

letFunny commented on internal/expr/parser.go:
```diff
@@ -942,6 +970,59 @@ func (p *Parser) parseColumnsInsertExpr() (expression, bool, error) {
+}
+
+// parseBasicInsertValues parses insert expressions that are value on the
+// right hand side of a basic insert expression.
```
Ditto for `basic insert expression`.

letFunny commented on internal/expr/parser_helper_test.go:
```diff
@@ -306,3 +306,49 @@ func (s parseSuite) TestAdvanceToNextExpr(c *C) {
+	}
+}
+
+func (s parseSuite) TestParseRenamingInsertValuesFalse(c *C) {
```
Should we also cover the corner cases like `()`, `("literalNoSpaces")`? These proved troublesome in the past.

Aflynn50 replied:
Added

letFunny commented on internal/typeinfo/valuelocator.go:
```diff
@@ -65,20 +88,49 @@ func (mk *mapKey) ArgType() reflect.Type {
+		}
+		argType = m.Type()
+		vals = append(vals, v.Interface())
+	} else if ms, ok := locateBulkType(typeToValue, mk.mapType); ok {
```
[Nit]:
> In general prefer early returns and separate conditionals

there is no need for `bulk` if you return early.

letFunny commented on internal/typeinfo/valuelocator.go:
```diff
@@ -131,19 +183,55 @@ func (f *structField) ArgType() reflect.Type {
+// value of this field.
+func (f *structField) LocateParams(typeToValue TypeToValue) (*Params, error) {
+	omit := false
+	bulk := false
```
[Nit]: Same nit about early returns and flags, that would avoid one level of conditionals and make it more explicit.

letFunny commented on internal/typeinfo/valuelocator.go:
```diff
@@ -131,19 +183,55 @@ func (f *structField) ArgType() reflect.Type {
+				if i == 0 && val.IsZero() {
+					omit = true
+				} else if val.IsZero() != omit {
+					return nil, fmt.Errorf("got mix of zero and none zero values in %s which has the omitempty flag set, in a bulk insert, values must be all zero or all none zero", f.Desc())
```
This test case is missing.

letFunny reviewed (commented):
One test file to go, I have reviewed the rest in depth. Changes look very good and the architecture and organization looks great. Thanks!

I have added some comments mainly about nitpicks, test cases and small improvements to readability. I don't want to block the PR so I will approve once I review the last file.

letFunny commented on internal/typeinfo/valuelocator_test.go:
```diff
@@ -127,162 +127,266 @@ func (s *typeInfoSuite) TestLocateScanTargetError(c *C) {
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "foo")
+		},
+		err: `got nil pointer in slice of "M" at position 1`,
```
Should we say `index` instead of position? Index to me starts at 0 clearly.

Aflynn50 replied:
Changed

letFunny commented on internal/typeinfo/valuelocator_test.go:
```diff
@@ -127,162 +127,266 @@ func (s *typeInfoSuite) TestLocateScanTargetError(c *C) {
-	argInfo, err := GenerateArgInfo([]any{S{}, T{}})
-	c.Assert(err, IsNil)
-
+func (s *typeInfoSuite) TestLocateParams(c *C) {
```
Nice refactor!

letFunny reviewed (approved):
Thanks

Aflynn50 committed:
Address further review comments

Aflynn50 commented:
> Out of curiosity more than anything. From the PR description, the SQL sent to the driver is:
> 
> ```
> INSERT INTO  t (col1, col2, col3, col4) VALUES (@sqlair_0, @sqlair_1, "v3", @sqalir_2), (@sqlair_3, @sqlair_4, "v3", @sqalir_5), (@sqlair_6, @sqlair_7, "v3", @sqalir_8)
> ```
> 
> Yet, `col2`, will always be a map for a constant value, so the query _could_ be turned into?
> 
> ```
> INSERT INTO  t (col1, col2, col3, col4) VALUES (@sqlair_0, @sqlair_1, "v3", @sqalir_2), (@sqlair_3, @sqlair_1, "v3", @sqalir_4), (@sqlair_5, @sqlair_1, "v3", @sqalir_6)
> ```

Thats a good point, no reason not to do this. I've changed the code and updated the PR description to reflect this.

Aflynn50 commented:
> I'd love to see a full end to end test added to [package_test.go](https://github.com/canonical/sqlair/blob/main/package_test.go) that baked in the bulk insert changes.
> 
> I think the more common approach that I regularly want to use bulk insert in sqlair for is the following:
> 
> ```
> type S struct {
>     a string `db:"col1"`
>     b string `db:"col2"`
>     c string `db:"col3"`
>     d string `db:"col4"`
> }
> 
> s := sqlair.MustPrepare(
>     "INSERT INTO t (col1, col2, col3, col4) VALUES ($S.*)",
>     S{},
> )
> 
> db.Query(
>     s,
>     []S{
>         {a: "a1", b: "b1", c: "c1", d: "d1"}, 
>         {a: "a2", b: "b2", c: "c2", d: "d2"},
>     },
> )
> ```
> 
> Name 	Value
> sqlair_0 	"a1"
> sqlair_1 	"b1"
> sqlair_2 	"c1"
> sqlair_3 	"d1"
> sqlair_4 	"a2"
> sqlair_5 	"b2"
> sqlair_6 	"c2"
> sqlair_7 	"d2"

Good point, I'd been meaning to add something to package_test but had forgotten. I've added a test now that has a simple case and a more complex case of a bulk insert.

Aflynn50 committed:
Add visitor pattern to bindinputs.go

manadart commented on internal/expr/bindinputs.go:
```diff
@@ -151,10 +106,121 @@ func newInsertColumn(input typeinfo.Input, column string, explicit bool) insertC
 	}
 }
 
+// bindInputs generates and verifies the query parameters corresponding to the
```
Just above here on line 95, there's a typo "explicity".

manadart commented on internal/expr/bindinputs.go:
```diff
@@ -126,13 +65,29 @@ func newOutputColumn(tableName string, columnName string, output typeinfo.Output
 	return outputColumn{column: tableName + "." + columnName, output: output}
 }
 
+type typedExpr interface {
```
I'd locate this at the top, before it's first usage.

manadart commented on internal/expr/bindinputs.go:
```diff
@@ -126,13 +65,29 @@ func newOutputColumn(tableName string, columnName string, output typeinfo.Output
-// insertColumn stores information about a single column of a row in an INSERT
+// addToQueryBuilder adds the typed input expressions to the query builder.
+func (te *typedInputExpr) addToQueryBuilder(qb *queryBuilder) error {
+	return qb.addInput(te)
```
This is not really _visitor_, because `queryBuilder` holds all the logic, and if you add another `typedExpr` implementation that it needs to handle, it grows.

That said, this _is_ managing complexity after a fashion, though you've tightly coupled the builder and the expressions.

manadart replied:
One way to do it would be for `queryBuilder` to implement an interface that exposes what the visitors need, then push the logic back into the visitors. Looking at:
```go
func (qb *queryBuilder) addInput(te *typedInputExpr) error {
	params, err := te.input.LocateParams(qb.typeToValue)
	if err != nil {
		return err
	}
	if params.Omit {
		return omitEmptyInputError(te.input.Desc())
	}
	if params.Bulk {
		return fmt.Errorf("cannot use bulk inputs outside an insert statement")
	}

	qb.argUsed[params.ArgTypeUsed] = true
	qb.sqlBuilder.writeInput(qb.inputCount, len(params.Vals))
	for _, val := range params.Vals {
		namedInput := sql.Named("sqlair_"+strconv.Itoa(qb.inputCount), val)
		qb.namedInputs = append(qb.namedInputs, namedInput)
		qb.inputCount++
	}
	return nil
}
```
The first half is validation that could (and should) be on `typedInputExpr`.
The rest could be done by coordinating methods like `setArgTypeUsed`, `writeInput`, `addNamedInput` or some such.

manadart reviewed (approved):
This is worth landing to get the feature. 

I would advocate thinking about how to refactor the `queryBuilder` interactions. I'll ruminate on this myself.

Aflynn50 committed:
Refactor logic out of query builder

Aflynn50 committed:
Add inputAssinger

letFunny commented on internal/expr/querybuilder.go:
```diff
@@ -122,16 +114,28 @@ func (qb *queryBuilder) checkAllArgsUsed(typeToValue typeinfo.TypeToValue) error
+// inputAssigner assigns input numbers to input expressions.
+type inputAssigner struct {
+	inputCount int
+}
```
To me this looks like a code smell, that probably there is too much abstraction for the given complexity. In my opinion, what you had at https://github.com/canonical/sqlair/pull/145/commits/afade243d0d0a68049dccfe4167e009272f92f9a was great, it moved the logic to encapsulate it and made the code clearer and more self-contained.

We now seem to be marching towards more encapsulation and decoupling, which is a worthwhile goal don't get me wrong. However, to me the results look a little bit too complex. Maybe we could wait until we need to add something else to the code and then do the refactor once we have new concrete use-cases.

Aflynn50 replied:
As discussed in person, we'll agree to disagree. In my opinion this is not adding complexity, it is making explicit complexity that is already there. The `inputAssigner` allows the `bindInputs` method on the `insertColumn` to take care of assigning the input numbers for the `boundInsertColumn` to use. Before it was being done by the caller based on the return values which (in a very minor way) broke the encapsulation of the the `boundColumn`.

This also makes changes to the way inputs are assigned a lot easier. We just change the `inputAssigner` struct and then follow through the compiler errors where `assignInputs` is called.

Aflynn50 committed:
Add more comments to inputAssigner
